### PR TITLE
[ShaderGraph] [Backport 10.3] Fix texture associated state (samplerstate, TexelSize) for subgraphs, CFNs

### PR DESCRIPTION
**Pull request by cdxntchou**
### Purpose of this PR

Backport to 10.x.x/release of https://github.com/Unity-Technologies/Graphics/pull/2651

Fix for https://fogbugz.unity3d.com/f/cases/1284036/
And https://fogbugz.unity3d.com/f/cases/1292031/

This PR converts the shadergraph resource-types (Texture2D, Texture2DArray, Texture3D, TextureCube, and SamplerState) to use a struct, instead of bunch of loosely associated variables.  This simplifies a bunch of the argument passing, and fixes a number of issues in the generated code, where we assume we can find the associated variables by name-munging, which often does not work.  This also clears up an issue with GLES2 where any explicit SamplerState used anywhere in the shader would cause it not to compile on the GLES2 platform (it will now compile, but will ignore the explicit SamplerStates and effectively always use the texture-associated sampler state).

We specially handle existing Custom Function Nodes that input or output one of these resource types, by introducing the concept of "Bare" resource types for Custom Function Nodes.  The Bare resource types represent JUST the resource itself, not the new resource struct.  Existing Custom Function Nodes will be versioned to use Bare resource types, so that the users existing HLSL code continues to work.  And any CFN using bare resource types will flag a warning to the user that those nodes should be upgraded to use the resource structs instead, guiding the users away from using Bare resource types (as the Bare resource types have all of the bug issues we are solving with the structs).

---
### Testing status
Tested URP (DX11) with all nodes that input or output one of the affected types:

- Texture2D Asset Node, Texture2D Input slot, Texture2D property type
  - SampleTexture2D -- works
  - SampleTexture2DLOD nodes -- works
  - TexelSize node -- works
  - NormalFromTexture node -- works
  - ParallaxMapping node -- works
  - ParallaxOcclusionMapping node -- works
  - Triplanar node -- works
  - HDRP Thread Map Detail (subgraph) -- works
  - Also all of the above work with a Texture2D value passed into and out of a SubGraph

image

- Cubemap Asset node, Cubemap input slot, Cubemap property type:
  - SampleCubemap (reflected) node works
  - SampleRawCubemap node works
  - Cubemap value passed into and out of a SubGraph

image

- Texture2DArray Asset node, input slot, property type
  - SampleTexture2DArray node works
  - Texture2DArray value passed into and out of a SubGraph works

image

- Texture3D Asset node, input slot, property type
  - SampleTexture3D node works
  - Texture3D value passed into and out of a SubGraph works

image

Ran same tests as above with HDRP (DX11)
Ran same tests as above with HDRP (Vullkan)
Ran same tests as above with URP (OpenGL Core 4.5)
Ran same tests as above with URP (DX12)
Ran same tests as above with URP (GLES2) (which has no samplers)
  Verified that using non-associated sampler states will now compile (yay)!   just ignored in the results
  NormalFromTexture node now works (previously did not compile in GLES2)
  CustomFunctionNodes taking SamplerState now work (in most cases)

  Broken nodes in GLES2: 
      ParallaxOcclusionMapping  (also broken in GLES2 before -- because looping not supported)
      CustomFunctionNodes of type File, taking SamplerState inputs, will not compile until the user converts the SamplerState type to use the new UnitySamplerState struct.   (type String works, as we fix the type automatically).
      SampleTexture3D fails (error C7532: global type sampler3D requires "#version 300" or later)
      SampleTexture2DArray fails (#error "SAMPLE_TEXTURE2D_ARRAY" is not supported on GLES 2.0)

Tested that an old CustomFunctionNode that used to take a Texture2D input is upgraded to take a BareTexture2D input, and continues working.  Same for Texture3D, Texture2DArray, TextureCube and SamplerState inputs and outputs:

image

image

Tested that an old CustomFunctionNode that used to return a Texture2D output, is upgraded to convert the Bare Texture2D output to a UnityTexture2D struct, and continues working (at least in all cases when it used to work -- i.e. when explicit sampler states are provided).  It also now generates a warning on the CustomFunction node:

image

Validated that converting the CustomFunctionNodes to use struct-based resources results in operational nodes with no warnings, and that it continues to work after save / load (versioning is applied correctly).

image

Tested VFX path can use texture sampling:
image

Tested raytracing locally -- appears to work with ShaderGraph materials.
Running the Raytracing TestProject locally on the ShaderGraph material tests pass.

Found issues using non-texture associated SamplerStates with VFX graphs (you get undefined symbol on the explicit sampler state), but appears to be a pre-existing issue (repros before this PR as well).

Another issue is that updating a ShaderGraph (via save/reimport) does not automatically propagate to VFX using that ShaderGraph.  Even re-importing the VFX doesn't do it.  You have to load the VFX into VFX graph, and save it in order to get the ShaderGraph changes.  Also seems to be a pre-existing issue that repros before this PR.

ABV:  🟢 
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/10.x.x%252Fsg%252Ffix%252Ftextures/.yamato%252F_abv.yml%2523all_project_ci_2020.2/4349472/job/pipeline

Console Tests:  🟢  
https://yamato.cds.internal.unity3d.com/jobs/339-ScriptableRenderPipelinePrivate/tree/10.x.x%252Fsg%252Ffix%252Ftextures2/.yamato%252Fall-shadergraph.yml%2523All_ShaderGraph_2020.2/4347732/job/pipeline

10.x.x/release ABV: (386699d)
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/10.x.x%252Frelease/.yamato%252F_abv.yml%2523all_project_ci_2020.2/4309391/job/pipeline

Failures:
   Only failing tests are the package-building ones, which only fail because this PR uses version 10.3 in the changelog, and it is expecting the 10.2 section to be first.

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

**Review by zeroyao — APPROVED** [review]
Thank you Chris!

**Review by LandonTownsendUnity — APPROVED** [review]
Things tested:
Inline on subgraph input 
image
subgraph property passthrough
image
Inline to subgraph output
image
inline node to subgraph output
image
"none" texture output from subgraph (make sure there's no errors)

All of the above for:
Texture2D
Texture2DArray
Texture3D
Cubemap

**Review by pbbastian — APPROVED** [review]
Lovely change! 👍 Only 1 minor change request

File: com.unity.shadergraph/Editor/Data/Nodes/Utility/CustomFunctionNode.cs [hunk 1]
```diff
@@ -103,50 +108,80 @@ public string functionBody
...
-                if (!first)
-                    call += ", ";
-                first = false;
-                call += GetVariableNameForSlot(argument.id);
+                    // fixup input for Bare texture types
+                    if (argument.bareResource)
+                        if (argument is SamplerStateMaterialSlot)
+                            sb.Append(".samplerstate");
+                        else
+                            sb.Append(".tex");
```
Comment (pbbastian) [inline com.unity.shadergraph/Editor/Data/Nodes/Utility/CustomFunctionNode.cs:158]: I really think this should have braces

**Review by pbbastian — CHANGES_REQUESTED** [review]
missed some minor things :)

File: com.unity.shadergraph/Editor/Data/Graphs/CubemapMaterialSlot.cs [hunk 2]
```diff
@@ -19,6 +20,22 @@ public CubemapMaterialSlot()
             : base(slotId, displayName, shaderOutputName, slotType, stageCapability, hidden)
         {}
 
+        [SerializeField]
+        internal bool m_BareResource = false;
```
[the remark this answers is not in the document]
Comment (cdxntchou) [reply com.unity.shadergraph/Editor/Data/Graphs/CubemapMaterialSlot.cs:24]: Fixed in new changeset

File: com.unity.shadergraph/Editor/Data/Nodes/Utility/CustomFunctionNode.cs [hunk 3]
```diff
@@ -103,50 +108,80 @@ public string functionBody
...
+                sb.Append(");");
+                sb.AppendNewLine();
+
+                foreach (var output in outputSlots)
+                {
+                    if (output.bareResource)
+                    {
+                        // fill in other values - sampler state, etc ??
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (cdxntchou) [reply com.unity.shadergraph/Editor/Data/Nodes/Utility/CustomFunctionNode.cs:183]: Yes, fixed in the new changeset -- For bare outputs I initialize the default output state above in the loop that declares the output variables.

File: com.unity.shadergraph/Editor/Data/Nodes/Utility/CustomFunctionNode.cs (hunk 1 above)
Comment (cdxntchou) [reply com.unity.shadergraph/Editor/Data/Nodes/Utility/CustomFunctionNode.cs:158]: Fixed in new changeset

**Review by pbbastian — APPROVED** [review]

**Review by alelievr — APPROVED** [review]
LGTM (for HDRP)

**Review by sebastienlagarde — APPROVED** [review]

**Review by elizabeth-legros — APPROVED** [review]

**Review by elizabeth-legros — APPROVED** [review]

**Review by VladNeykov — COMMENTED** [review]
A bit late to the party, but it's all good on the VFX side.
Ran FTP project, checked a VFX sample (Magic book) using shadergraph, and did a few manual tests.

**Pull request merged**